fix: CLI blogpost draft (#638)
ory/web: src/markdown/blog/ory-cloud-cli-devtool-examples.mdx

seo:
  title: |
    Ory CLI title
  description: |
    Ory CLI description
  keywords: |
    login, auth, authentication, registration, CLI, cloud native, open source, ory, ory kratos, command line interface, command line, bash, shell, script, gitflow, gitops, git, automation

publishedAt: '2022-05-08'
author: vinckr

overline: |
  Developer CLI
tags:
  - Guide
  - Authentication
  - Ory Cloud
  - Open Source
  - User Management
---

Great developer experience is the founding tenet of Ory and as such we are happy to now have a powerful CLI tool that lets you run and configure your Ory Cloud projects. As the CLI is under constant development, this series of blogposts aims to highlight some of the less documented or publicized aspects of the Ory CLI. Why am I so excited about the new Ory CLI? Because now you have the full configuration for Ory Kratos -  (Ory Kratos is the [open source identity service that is used by Ory Cloud](https://github.com/ory/kratos)) - at your fingertips. Previously we were limited in what was available through the [console UI](https://console.ory.sh/) - which to be honest covers most use cases - but now we can configure the rest through the Ory CLI.

> In this post I am going to show some thing you can do with `ory patch project`, which is the CLI command to change (amend, patch) the Ory Cloud Project Configuration. You can also [upload a full configuration using `ory update project`](https://www.ory.sh/docs/guides/cli/config-with-cli#overwrite--import-configuration) - or rather overwrite with your full configuration.  With [`ory patch projects`](https://www.ory.sh/docs/guides/cli/config-with-cli#patch-configuration) you just configure single items (e.g. a redirect or token lifetime), for bigger changes it is easier to use a full configuration with [`ory update project`](https://www.ory.sh/docs/guides/cli/config-with-cli#overwrite--import-configuration). With this you can automatically switch configurations between different projects, for example staging and production and I will cover it in-depth in the next edition of this series. 

Reviewer: ```suggestion
In this post I am going to show some things you can do with `ory patch project` and `ory update project`, which are the CLI commands to update or overwrite your Ory Cloud Project configurations. 

You can [upload a full configuration using `ory update project`](https://www.ory.sh/docs/guides/cli/config-with-cli#overwrite--import-configuration) - or rather overwrite with your full configuration.  With [`ory patch projects`](https://www.ory.sh/docs/guides/cli/config-with-cli#patch-configuration) you just configure single items (e.g. a redirect or token lifetime). 

Be cautious of importing the full configuration file with `ory update project` as it will **overwrite** your Ory Cloud Project configurations - essentially you could by accident delete some configuration values. When moving from Ory Kratos Open Source, it might be more beneficial to import the configuration with `ory update project`.
```

I think we should mention how dangerous it is to use `ory update project` since you can essentially remove some of the configuration values by accident (if you don't pass the full configuration file back to the CLI). You would only want to do something like this when moving from open source to cloud since you don't have any configuration file in Cloud and now you move over. 

You can also use this in an automation pipeline to for example version configuration files using Git. So when you need to roll back to a previous configuration file you can just pass the previous version to `ory update project`. 

I would also not recommend changing between production and staging etc using this configuration update approach - it would be better to run a couple of Ory Cloud projects.

Author: >With this you can automatically switch configurations between different projects, for example staging and production

>I would also not recommend changing between production and staging etc using this configuration update approach - it would be better to run a couple of Ory Cloud projects.

I meant that you can export the configuration from your staging project and upload it to your production project. Would you not recommend doing that?

Reviewer: no you definitely can, but will need to probably change up some of the redirect urls and domain setup :)

The base command is the following:

```bash
ory patch project <your-project-id> --replace '/services/identity/config/...'
```

> `/services/identity/config/`

Reviewer: I think we should say here that we have a couple of flags we could pass to `ory patch project`.

- `--replace`
- `--add`
- `--delete`
- `--format`

Only giving a singular command that is incomplete doesn't make sense in a guide. We need to show people how to do something step by step. For instance, you require the project id to make a change. How do you do that? Do we automatically assume the person has the CLI installed, a cloud account and a project setup?

We need to have a section at the start going through the _basic setup_.

Reviewer: Here is an example of the CLI printing out instructions on how to use the `ory patch project` command.

```shell
Usage:
  ory patch project id [flags]

Examples:
ory patch project ecaaa3cb-0730-4ee8-a6df-9553cdfeef89 \
	--replace '/name="My new project name"' \
	--add '/services/identity/config/courier/smtp={"from_name":"My new email name"}' \
	--replace '/services/identity/config/selfservice/methods/password/enabled=false' \
	--delete '/services/identity/config/selfservice/methods/totp/enabled'

ory patch project ecaaa3cb-0730-4ee8-a6df-9553cdfeef89 \
	--replace '/name="My new project name"' \
	--delete '/services/identity/config/selfservice/methods/totp/enabled'
	--format kratos-config > my-config.yaml
```

As we want to configure Ory Kratos, the Ory identity services, that is the standard path for the replace flag.
If we want to for example configure the UI URL of the self-service registration flow this is the path we would use.

> `/services/identity/config/selfservice/flows/registration/ui_url/`

Reviewer: I think we need a bit more than just the path. Maybe show how the configuration file is structured in `yaml` and `json` and then how that translates to the cli path.

[Have a look at the full Ory Kratos configuration](https://ory.sh/docs/kratos/reference/configuration) to get an idea of this is is structured.

## Export full configuration

If you want to check out your full Ory Cloud project configuration you can export it with a simple command 

```bash

> ory get identity-config <your-project-id> --format json-pretty

Reviewer: I think `identity-config` is an alias and we can also use `kratos-config` here. We should maybe just confirm which command is the latest. This is also only the `kratos` configurations (the config is converted so now you can use it in open source). 

You can use 

```shell
ory get kratos-config <project id> --format yaml > kratos.yaml
````

But if you wanted to get your project configurations you need to use `ory get project`.

```shell
ory get project <project id> --format json
```

Author: Both kratos-config and identity-config work, i think identity config is more intuitive, but we should mention both 🤔 

tbh at the moment its only really the kratos config, but that will probably change soon

---
path: "/cli-post/"

title: |
  Ory Devtools: Ory CLI `patch project`

teaser: |

>   Ory CLI teaser text

Reviewer 2: Can we please change the metas?

seo:
  title: |
    Ory CLI title
  description: |
    Ory CLI description
  keywords: |
    login, auth, authentication, registration, CLI, cloud native, open source, ory, ory kratos, command line interface, command line, bash, shell, script, gitflow, gitops, git, automation

publishedAt: "2022-05-08"
author: vinckr

overline: |
  Developer CLI
tags:
  - Guide
  - Authentication
  - Ory Cloud
  - Open Source
  - User Management
---

Great developer experience is the founding tenet of Ory and as such we are happy
to now have a powerful CLI tool that lets you run and configure your Ory Cloud
projects. As the CLI is under constant development, this series of blogposts
aims to highlight some of the less documented or publicized aspects of the Ory
CLI. Why am I so excited about the new Ory CLI? Because now you have the full
configuration for Ory Kratos - (Ory Kratos is the
[open source identity service that is used by Ory Cloud](https://github.com/ory/kratos)) -
at your fingertips. Previously we were limited in what was available through the
[console UI](https://console.ory.sh/) - which to be honest covers most use
cases - but now we can configure the rest through the Ory CLI.

> In this post I am going to show some thing you can do with `ory patch project`,

Reviewer 2: ```suggestion
In this post, I am going to show some things you can do with `ory patch project`,
```

which is the CLI command to change (amend, patch) the Ory Cloud Project
Configuration. You can also
[upload a full configuration using `ory update project`](https://www.ory.sh/docs/guides/cli/config-with-cli#overwrite--import-configuration) -
or rather overwrite with your full configuration. With
[`ory patch projects`](https://www.ory.sh/docs/guides/cli/config-with-cli#patch-configuration)
you just configure single items (e.g. a redirect or token lifetime), for bigger

> changes it is easier to use a full configuration with

Reviewer 2: ```suggestion
changes, it is easier to use a full configuration with
```

[`ory update project`](https://www.ory.sh/docs/guides/cli/config-with-cli#overwrite--import-configuration).
With this you can automatically switch configurations between different
projects, for example staging and production and I will cover it in-depth in the
next edition of this series.

The base command is the following:

```bash
ory patch project <your-project-id> --replace '/services/identity/config/...'
```

`/services/identity/config/`

As we want to configure Ory Kratos, the Ory identity services, that is the
standard path for the replace flag. If we want to for example configure the UI
URL of the self-service registration flow this is the path we would use.
`/services/identity/config/selfservice/flows/registration/ui_url/`

[Have a look at the full Ory Kratos configuration](https://ory.sh/docs/kratos/reference/configuration)

> to get an idea of this is is structured.

Reviewer 2: ```suggestion
to get an idea of this structured (what)?
```

## Export full configuration

> If you want to check out your full Ory Cloud project configuration you can

Reviewer 2: ```suggestion
If you want to check out your full Ory Cloud project configuration, you can
```

export it with a simple command

```bash
ory get identity-config <your-project-id> --format json-pretty
```

The normal output is more machine than human-readable, so I recommend you use
the flag `--format json-pretty` to format it nicely - or `--format yaml` if you

> prefer that ;) For example to confirm that everything is configured correctly -

Reviewer 2: ```suggestion
prefer that ;) For example, to confirm that everything is configured correctly,
```

> you can also do this on console.ory.sh but not all configuration options have an

Reviewer 2: ```suggestion
you can also do this on console.ory.sh, but not all configuration options have a
```

UI. You can also export your configuration from a staging or development project
and copy it over to your production project, or implement a git-based
configuration workflow using the Ory CLI.

> Let us know what usecases you can think of or would like to see supported with

Reviewer 2: I think this sentence could be placed at the end of the article. Makes more sense there.

## Custom UI configuration

```bash
ory patch project <your-project-id> --replace '/services/identity/config/selfservice/flows/registration/ui_url="http://localhost:3000/registration"'
```

```bash
ory patch project <your-project-id> --replace '/services/identity/config/selfservice/flows/login/ui_url="http://localhost:3000/login"'
```

```bash
ory patch project <your-project-id> --replace '/services/identity/config/selfservice/flows/settings/ui_url="http://localhost:3000/settings"'
```

```bash
ory patch project <your-project-id> --replace '/services/identity/config/selfservice/flows/verification/ui_url="http://localhost:3000/verification"'
```

```bash
ory patch project <your-project-id> --replace '/services/identity/config/selfservice/flows/recovery/ui_url="http://localhost:3000/recovery"'
```

```bash
ory patch project <your-project-id> --replace '/services/identity/config/selfservice/flows/error/ui_url="http://localhost:3000/error"'
```

You can also configure those through the console UI.

## Default Return URLs

We can also configure the URL where users end up after a selfservice flow, for
example after logging out of the application you usually want to display the

> login screen. So as before we use
> `/services/identity/config/selfservice/flows/logout/` as base config path but
> this time adding `after/default_browser_return_url` to configure what happens
> _after_ the logout.

Reviewer 2: ```suggestion
login screen. So as before, we use
`/services/identity/config/selfservice/flows/logout/` as base config path but
this time we add `after/default_browser_return_url` to configure what happens
_after_ the logout.
```

```bash
ory patch project <your-project-id> --replace '/services/identity/config/selfservice/flows/logout/after/default_browser_return_url="/ui/login"'
```

> In the same way we can configure the other selfservice flow return URLs, for

Reviewer 2: ```suggestion
In the same way, we can configure the other selfservice flow return URLs, for
```

example when you want the user to re-authenticate after changing the password in
the settings flow, you can configure this in `/flows/settings/after/password/`:

```bash
ory patch project <your-project-id> --replace '/services/identity/config/selfservice/flows/settings/after/password/default_browser_return_url="/ui/login"'
```

You can also configure these return URLs through the console UI.

## Identity Schemas

The identity schema can be most easily configured in the console UI, but you can
still do some things through the CLI if you prefer. This is one of the areas

> where I would actually much prefer the console UI with validation, naming the

Reviewer 2: ```suggestion
where I would prefer the console UI with validation, naming the
```

> schemas and so on. Most of the time you wont make changes to the identity schema

Reviewer 2: ```suggestion
schemas and so on. Most of the time, you won't make changes to the identity schema
```

often and save buiness-logic related data outside of Ory. With the export

> function you can find out the id and url of your current schema. If you are

Reviewer 2: ```suggestion
function, you can find out the id and url of your current schema. If you are
```

> using for example the default email/password schema the id is `preset://email`,

Reviewer 2: ```suggestion
using the default email/password schema, for example, the id is `preset://email`,
```

> but for a custom schema the id will be a long string. We can change the default

Reviewer 2: ```suggestion
but for a custom schema, the id will be a long string. We can change the default
```
